File: solutions/clojure/pythagorean-triplet/1/src/pythagorean_triplet.clj
(ns pythagorean-triplet)

(defn make-triplet [n a]
     (let [na (- n a)
           numerator (+ (* a a) (* na na))
           denominator (* 2 na)]
         (when (zero? (mod numerator denominator)) 
           (let [c (/ numerator denominator)
                 b (- n c a)]
               (when (> b a)
                 (list a b c))
             ))))

(defn find-pythagorean-triplets
  "Given an integer N, it returns all Pythagorean triplets for which a + b + c = N"
  [N]
  (filter some? (map #(make-triplet N %) (range 1 (quot N 3))))
  )

File: solutions/javascript/pythagorean-triplet/1/pythagorean-triplet.js
//
// pythagorean triplets 
//

export function triplets({ minFactor, maxFactor, sum }) {
  let res = []
  const [min, max] = [minFactor ?? 1, maxFactor ?? sum/2]
  for (let a=min; a<sum/3; a++){
    let numerator = a*a + (sum-a)*(sum-a)
    let denominator = 2*(sum-a)
    if(0==numerator%denominator){
      let c = numerator/denominator
      let b = sum - a - c
      if(b>a && c<=max)
        res.push(new Triplet(a,b,c))
    }
  }
  return res  
}

class Triplet {
  constructor(a, b, c) {
    this.triplet = [a,b,c]
  }

  isPythagorean(){
    const [a,b,c] = this.triplet
    return a<b && b<c && a*a + b*b == c*c
  }

  toArray() {
   return this.triplet
  }
}
